### DISCUSSION TITLE
Arguments in MMFF Constraints

### DISCUSSION THREAD

**Discussion by Nokimann** [rdkit/rdkit#5170] [Q&A] [answered]
I'm generating a 3D molecule structure with the MMFF scheme. I have two questions.

1. There exist an `relative` argument in `MMFFAddAngleConstraint`, `MMFFAddDistanceConstraint`, and `MMFFAddTorsionConstraint`.
But, I cannot understand how to work the `min*` and the `max*` arguments with the `relative=True` argument.

How can I set the `min*` and the `max*` values? Do they have to be positive? Or `min*` must be negative and `max*` be positive?
Does it works like a constraint satisfying ```initial + min_constraint (negative value) < MMFF optimized value < initial + max_constraint (positive value)```

2. If we set  `relative=False`, what is the range of angle and torsion?

angle ~ [0, 180] / torsion ~[-180, 180], [0, 360] ... ?

**Comment by ptosco** [comment, marked as answer]
@Nokimann 

Regarding point (1)
```
from rdkit import Chem
from rdkit.Chem import AllChem
import matplotlib.pyplot as plt

m = Chem.AddHs(Chem.MolFromSmiles("CCCC"))
AllChem.EmbedMolecule(m)
0

mp = AllChem.MMFFGetMoleculeProperties(m)
globalFF = AllChem.MMFFGetMoleculeForceField(m, mp)
m2 = Chem.Mol(m)
y = []
x_values = list(range(-180, 180, 5))
for x in x_values:
    ff2 = AllChem.MMFFGetMoleculeForceField(m2, mp)
    AllChem.SetDihedralDeg(m2.GetConformer(), 0, 1, 2, 3, x)
    ff2.MMFFAddTorsionConstraint(0, 1, 2, 3, False, x - .1, x + .1, 10000.0)
    ff2.Minimize()
    y.append(globalFF.CalcEnergy(ff2.Positions()))

plt.plot(x_values, y)
```
![image](https://user-images.githubusercontent.com/5244385/162004204-82e8d3db-c3b5-4c06-801e-79865ba7aa1c.png)

`min` and `max` when `relative=True` are interpreted as deltas from the current value.
The most obvious way to use them is to specify a negative value for `min` and a positive value for `max`:

```
AllChem.SetDihedralDeg(m2.GetConformer(), 0, 1, 2, 3, 0)
ff = AllChem.MMFFGetMoleculeForceField(m2, mp)
ff.MMFFAddTorsionConstraint(0, 1, 2, 3, True, -10, +10, 10000.0)
ff.Minimize(maxIts=10000)
0
AllChem.GetDihedralDeg(m2.GetConformer(), 0, 1, 2, 3)
10.000002970915057
```

However, this is not a requirement:
```
AllChem.SetDihedralDeg(m2.GetConformer(), 0, 1, 2, 3, 0)
ff = AllChem.MMFFGetMoleculeForceField(m2, mp)
ff.MMFFAddTorsionConstraint(0, 1, 2, 3, True, -20, -10, 10000.0)
ff.Minimize(maxIts=10000)
0
AllChem.GetDihedralDeg(m2.GetConformer(), 0, 1, 2, 3)
-20.00000484140093
```

The only constraint is that `min` must be < `max` or an exception will be thrown:
```
AllChem.SetDihedralDeg(m2.GetConformer(), 0, 1, 2, 3, 0)
ff = AllChem.MMFFGetMoleculeForceField(m2, mp)
ff.MMFFAddTorsionConstraint(0, 1, 2, 3, True, -10, -20, 10000.0)
ff.Minimize(maxIts=10000)
---------------------------------------------------------------------------
RuntimeError                              Traceback (most recent call last)
/tmp/ipykernel_261430/2234620847.py in <module>
      1 ff = AllChem.MMFFGetMoleculeForceField(m2, mp)
----> 2 ff.MMFFAddTorsionConstraint(0, 1, 2, 3, True, -10, -20, 10000.0)
      3 ff.Minimize(maxIts=10000)

RuntimeError: Pre-condition Violation
	minDihedralDeg must be <= maxDihedralDeg
	Violation occurred on line 31 in file Code/ForceField/MMFF/TorsionConstraint.cpp
	Failed Expression: !(minDihedralDeg > maxDihedralDeg)
	RDKIT: 2022.09.1pre
	BOOST: 1_74
```

Regarding point (2), use the range that allows you to specify the constraint that you need.
For example, if you wish to constrain an angle around 180, it will be convenient to specify a 170, 190 range:
```
AllChem.SetDihedralDeg(m2.GetConformer(), 0, 1, 2, 3, 160)
ff = AllChem.MMFFGetMoleculeForceField(m2, mp)
ff.MMFFAddTorsionConstraint(0, 1, 2, 3, False, 170, 190, 10000.0)
ff.Minimize(maxIts=10000)
0
AllChem.GetDihedralDeg(m2.GetConformer(), 0, 1, 2, 3)
-179.99997680909718
```
If you wish to constrain an angle around 0, it will be convenient to specify a -10, 10 range:
```
AllChem.SetDihedralDeg(m2.GetConformer(), 0, 1, 2, 3, 20)
ff = AllChem.MMFFGetMoleculeForceField(m2, mp)
ff.MMFFAddTorsionConstraint(0, 1, 2, 3, False, -10, 10, 10000.0)
ff.Minimize(maxIts=10000)
0
AllChem.GetDihedralDeg(m2.GetConformer(), 0, 1, 2, 3)
9.999971471095357
```

**Reply by Nokimann** [reply]
Thank you,  @ptosco, for the specific examples.

In point (1), it seems that the final dihedral angle results in `(initial - 20)`  from the bound: `initial + (-20) < MMFF optimized value < initial + (-10)` when the arguments are `(min, max) = (-20, -10)` and `relative=True`

**Reply by ptosco** [reply]
@Nokimann That's because when I set the initial dihedral to 0 I generated a totally eclipsed conformation with high strain due to the steric clashes between the distal methyls. The molecule would very much like to get away from that conformation, but since the offending dihedral value is constrained to the range (-20, -10), it goes to -20, which is the farthest allowed from 0.
